### PR TITLE
Checkout: automate payment processor by standardizing responses

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This is a change to the `composite-checkout` package that automates every payment processor so that they are no longer responsible for setting the transaction status manually. They can each now just call the `onClick` handler (passed as a prop to the button), and the payment processor function will take care of the work of converting the submitted data into a standardized response. That response will then trigger transaction status changes automatically.

This requires the payment processor functions to take on any additional processing required to transform the actual server response into one of a set of standard responses that the `composite-checkout` package can handle. Currently there are three such responses represented by the returned Promise resolving with one of three codes: SUCCESS, REDIRECT, and NOOP. There's also the failure state which is represented by the Promise rejecting. SUCCESS and REDIRECT were already mostly handled by the transaction status system but had to be set manually; now the response will do that. The NOOP response allows the payment method to take its own actions if it wishes (eg: the WeChat payment method sometimes displays a QR code as a final state and handles its own redirect and error states).

Example of a payment method button updated to use the new system where you can see the removed boilerplate:

Screen Shot 2020-10-10 at 12 39 37 AM

This greatly simplifies making a payment method, although the amount of processing the server response is the same; the work is just moved to the payment processor function with the transaction status boilerplate removed.

In terms of the code, this also will make it much easier in the future to convert the payment method objects to TypeScript as it removes their most complex part. It also allows us to standardize on several patterns around redirects and error messages. Altogether these changes should lead to more reliable and flexible code with fewer bugs.

This PR only changes two payment methods: credit card (which actually includes the built-in credit card method used only for the demo, the stripe credit card method, and the ebanx method), and existing cards. Other payment methods will be converted in future PRs:

- PayPal: https://github.com/Automattic/wp-calypso/pull/46431
- Free and Full Credits: https://github.com/Automattic/wp-calypso/pull/46430
- Apple Pay: https://github.com/Automattic/wp-calypso/pull/46429
- Stripe Redirect methods: https://github.com/Automattic/wp-calypso/pull/46428

#### Testing instructions

- Begin a payment using PayPal and verify that once you click the submit button, you are redirected to PayPal (it should not be changed by this PR so no need to test the whole purchase).
- Complete a payment using a credit card processed through Stripe.
  - Use a [stripe test card](https://stripe.com/docs/testing).
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
- Complete a payment using a credit card processed through Stripe that requires 3DS.
  - Use a [stripe test card](https://stripe.com/docs/testing) that requires 3D Secure 1 or 2.
  - Verify that the 3DS modal is displayed, that the payment succeeds, and that you are redirected to the thank-you page.
- Complete a payment using an existing credit card.
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
- Complete a payment using an existing credit card that requires 3DS.
  - Verify that the 3DS modal is displayed, that the payment succeeds, and that you are redirected to the thank-you page.
- Complete a payment using a credit card processed through Ebanx.
  - Apply D48147-code to your sandbox to pretend to be in Brazil and sandbox the API.
  - Set your currency to `BRL`.
  - In the contact information step, choose "Brazil" for the country.
  - In the final step, choose "Credit or debit card".
  - Verify that the form asks for all the Ebanx fields in addition to the credit card fields (Taxpayer Identification Number, Phone, etc.)
  - Use an [ebanx test card](https://developer.ebanx.com/docs/resources/testCreditCards/) and [test customer info](https://developer.ebanx.com/docs/resources/testCustomerData/).
  - Verify that the payment succeeds and that you are redirected to the thank-you page.

### CODE REVIEW

File: packages/composite-checkout/src/components/checkout-submit-button.tsx [hunk 1]
```diff
@@ -1,21 +1,79 @@
...
+	const {
+		setTransactionComplete,
+		setTransactionRedirecting,
+		setTransactionError,
+		setTransactionPending,
+	} = useTransactionStatus();
+	const paymentProcessors = usePaymentProcessors();
+	const { __ } = useI18n();
+	const redirectErrorMessage = __(
+		'There was an error redirecting you to the payment partner. Please try again or contact support.'
```
Comment (a8ci18n) [inline packages/composite-checkout/src/components/checkout-submit-button.tsx:33]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'An error occurred while redirecting to the payment partner. Please try again or contact support.' )` *ES Score: 7*

**Comment by nbloomf** [comment]
Nice! This has been an outstanding cleanup item for a while. Here are the results of my testing:

- New stripe card, non-3DS, charge failure: notification is displayed and checkout reenters active state
- New stripe card, non-3DS, charge success: (personal plan) charge succeeds and calypso redirects to a thank-you page
- New stripe card, 3DS, auth failure: correct notification is displayed and checkout reenters active state
- New stripe card, 3DS, auth success, charge failure: correct notification is displayed and checkout reenters active state
- New stripe card, 3DS, auth success, charge success: (personal plan) charge succeeds and calypso redirects to a thank-you page
- Existing stripe card, non-3DS: (premium plan) charge succeeds and calypso redirects to an upsell page
- Existing stripe card, 3DS, auth failure: correct notification is displayed and checkout reenters active state
- Existing stripe card, 3DS, auth success: (personal plan) charge succeeds and calypso redirects to a thank-you page
- New ebanx card, charge failure: extra fields load as expected, correct notification is displayed, and checkout reenters active state. (Note- we should eventually make the ebanx errors a little nicer, I got `PaymentFailureError: The transaction was declined. Please try another card.`)
- New ebanx card, charge success: (premium) charge succeeds and calypso redirects to an upsell page
- Existing ebanx card: (business) charge succeeds and calypso redirects to a thank you page

TL;DR: no bugs!

I did notice something interesting though. When I used a stripe card number triggering a charge failure, we didn't save the card details, but when doing the same with ebanx, we did.

**Review by nbloomf — APPROVED** [review]

**Comment by sirbrillig** [comment]
Thank you for your amazing testing, as usual!

> I did notice something interesting though. When I used a stripe card number triggering a charge failure, we didn't save the card details, but when doing the same with ebanx, we did.

🤔  I imagine that has more to do with the Stripe elements (since with ebanx, they are our own fields) than with the payment method, but if you think it's worth investigating further, we could create an issue.

**Comment by sirbrillig** [comment]
Interesting, one of the failing e2e tests is 

```Viewing a specific plan with coupon: @parallel @jetpack
05:21:18          Can Correctly Apply Coupon:
05:21:18      TimeoutError: Timed out waiting for element with css selector of '.checkout-steps__step-complete-content .wp-checkout-order-review__show-coupon-field-button' to be clickable
05:21:18 Wait timed out after 22176ms```

and the screenshot looks like a coupon was applied but doesn't show up in the review step:

![Screen Shot 2020-10-14 at 4 26 24 PM](https://user-images.githubusercontent.com/2036909/96041818-6cbab700-0e3a-11eb-89b7-05b8f475d059.png)

This seems like if it _is_ a bug, it's not related to this PR (coupons are a shopping cart behavior, not having to do with payment methods or transaction status), but I'll just do a little research to see if I can learn anything.

**Comment by sirbrillig** [comment]
I verified that the failing mobile e2e test is because the coupon savings differs by $0.01 USD due to a rounding issue where the test is rounding down and the shopping cart is rounding up. I'd say that's not an issue.

I'm not able to reproduce the failing desktop e2e test with the coupon that I mentioned in the above comment, so I'm going to assume it was a fluke or an unrelated bug.
